src/library/scanner/libraryscanner.cpp

```diff
@@ -476,6 +510,58 @@ void LibraryScanner::addNewTrack(TrackPointer pTrack) {
         emit(trackAdded(pTrack));
         emit(progressLoading(pTrack->getLocation()));
     } else {
-        qWarning() << "Track ("+pTrack->getLocation()+") could not be added";
+        qWarning()
+                << "Track (" + pTrack->getLocation() + ") could not be added";
+    }
+}
+
+bool LibraryScanner::changeScannerState(ScannerState newState) {
+    switch (newState) {
+    case IDLE:
+        // we are leaving STARTING  or CANCELING state
+        // m_state is already IDLE if a scan was canceled
+        m_state = IDLE;
+        m_stateMutex.unlock();
+        return true;
+    case STARTING:
+        // we need to lock the mutex during the STARTING state
+        // to prevent loosing cancel commands or start the scanner
+        // twice
+        if (m_stateMutex.tryLock()) {
+            if (m_state != IDLE) {
+                qDebug() << "LibraryScanner: Scan already in progress.";
+                m_stateMutex.unlock();
+                return false;
+            }
+            m_state = STARTING;
+            return true;
+        } else {
+            qDebug() << "LibraryScanner: mutex locked, state =" << m_state;
+            return false;
+        }
+    case SCANNING:
+        DEBUG_ASSERT(m_state == STARTING);
+        // Transition protected by the mutex is over now
+        // Allow canceling
+        m_state = SCANNING;
+        m_stateMutex.unlock();
+        return true;
+    case CANCELING:
+        // canceling is always possible, but wait
+        // until there is no scan starting.
+        m_stateMutex.lock();
```

Comment:
> Would it be possible to change the state earlier to SCANNING or IDLE depending on dirs.isEmpty() at the beginning of this function?

Good point. Yes this should work since we don't depend on the value of m_scannerGlobal any more.